### DISCUSSION TITLE
Using ParamConverterProvider on client side with annotations

### DISCUSSION THREAD

**Discussion by itineric** [resteasy/resteasy#3666] [Q&A]
I am trying to use `ParamConverterProvider.getConverter(Class<T> rawType, Type genericType, Annotation[] annotations)` with RestEasy client but I am always getting "null" as annotations array.

The behaviour is not the same on server side.

I would like to factorize all converters by using annotations but it does not seem possible.

**Comment by jamezp** [comment]
Can you show an example? What version of RESTEasy are you using?

**Comment by itineric** [comment]
Sure I can. I simplified my usecase just to show the issue.

The service that is shared between client and server
```
@Path("test")
public interface TestResource
{
  @GET
  @Produces(MediaType.APPLICATION_JSON)
  void test(@QueryParam("obj") @SomeAnnotation SomeObject obj);
}
```

The server implementation (irrelevant for the current problem)
```
public class TestResourceImpl implements TestResource
{
  @Override
  public void test(final SomeObject obj)
  {
  }
}
```

The client

```
final ResteasyClient client = (ResteasyClient) ClientBuilder.newClient();
client.register(new SomeObjectParamConverterProvider());
final ResteasyWebTarget target = client.target(URI);

final TestResource testResource = target.proxy(TestResource.class);
final SomeObject someObject = new SomeObject();
testResource.test(someObject);
```

And of course the SomeObjectParamConverterProvider (also shared between client and server)
```
@Provider
public class SomeObjectParamConverterProvider implements ParamConverterProvider
{
  private static class SomeObjectParamConverter implements ParamConverter<SomeObject>
  {

    @Override
    public SomeObject fromString(final String param)
    {
      return new SomeObject();
    }

    @Override
    public String toString(final SomeObject someObject)
    {
      return "";
    }
  }

  @Override
  public <T> ParamConverter<T> getConverter(final Class<T> rawType,
                                            final Type genericType,
                                            final Annotation[] annotations)
  {
    System.out.println("Must convert " + rawType);
    System.out.println("Getting annotations " + Arrays.toString(annotations));
    if (rawType.isAssignableFrom(SomeObject.class))
    {
      @SuppressWarnings("unchecked")
      final ParamConverter<T> tmp = (ParamConverter<T>) new SomeObjectParamConverter();
      return tmp;
    }
    return null;
  }
}
```

On the server side I am getting:
```
Must convert class com.example.SomeObject
Getting annotations [@jakarta.ws.rs.QueryParam("obj"), @com.example.SomeAnnotation()]
```

On the client side I am getting
```
Must convert class com.example.SomeObject
Getting annotations null
```

I was expecting to get the annotations on the client side too, especially the "@SomeAnnotation" since the annotation is set on the common interface both client and server have access to.

**Comment by itineric** [comment]
After some digging, annotations are not passed through at this point: https://github.com/resteasy/resteasy/blob/28dd5465af7d399cb98d0837444f8f0bfda235bc/resteasy-client/src/main/java/org/jboss/resteasy/client/jaxrs/internal/proxy/processors/AbstractCollectionProcessor.java#L60
But they are available.

Continuing the stack of calls, it ends up calling https://github.com/resteasy/resteasy/blob/28dd5465af7d399cb98d0837444f8f0bfda235bc/resteasy-client/src/main/java/org/jboss/resteasy/client/jaxrs/internal/ClientConfiguration.java#L128
It should probably call https://github.com/resteasy/resteasy/blob/28dd5465af7d399cb98d0837444f8f0bfda235bc/resteasy-client/src/main/java/org/jboss/resteasy/client/jaxrs/internal/ClientConfiguration.java#L132

**Comment by itineric** [comment]
I think the AbstractCollectionProcessor.buildIt method should look more like this
```
public T buildIt(T target, Object object)
{
  if (object == null) return target;
  ParamConverter<Object> paramConverter = config.getParamConverter(object.getClass(), type, annotations);
  if (paramConverter != null)
  {
     object = paramConverter.toString(object);
     target = apply(target, object);
  }    
  else
  {
    if (object instanceof Collection)
    {
      target = apply(target,  ((Collection<?>) object).toArray());
    }
    else if (object.getClass().isArray())
    {
      Object[] arr = convertToObjectsArray(object);
      target = apply(target, arr);
    }
    else
    {
       target = apply(target, object);
    }
  }
  return target;
}
```

**Reply by jamezp** [reply]
I appreciate all this research. I want to let you know I'm not ignoring it, I just haven't had a chance to fully look at it. I likely won't be able to until next week.
